phylogeny/prototypeSelection.py

```diff
@@ -0,0 +1,106 @@
+"""
+Prototype selection
+
+Given a set of n elements (S) and their pairwise distances in terms of a
+dissimilarity matrix (DM) and a given number (k << n), find the sub-set (s)
+consisting of exactly k elements, such that s best represents the full set S.
+
+Here, we define "best represents" as maximizing the sum of distances between
+all points, i.e. sum {over a,b in S} DM[a,b]. This is our objective function.
+
+This problem is known to be NP-hard [1], thus we need to resort to heuristics.
+In the future, this module will implement several heuristics, whose quality can
+be measured by the objective function for each problem instance, since there is
+no global winner.
+For completeness, the exact but exponential algorithm is implemented, too.
+  "prototypeSelection_exhaustive"
+
+[1] Gamez, J. Esteban, François Modave, and Olga Kosheleva.
+    "Selecting the most representative sample is NP-hard:
+     Need for expert (fuzzy) knowledge."
+    Fuzzy Systems, 2008. FUZZ-IEEE 2008.
+"""
+
+from skbio.stats.distance import DistanceMatrix
+import itertools
+import numpy as np
+import scipy as sp
+
+
+def distanceSum(elements, dm: DistanceMatrix):
+    '''Compute the sum of pairwise distances for the given elements according to
+    the given dissimilarity matrix.
+
+    Parameters
+    ----------
+    elements: list or set
+        list or set of element IDs for which the sum of distances is computed
+    dm: skbio.stats.distance.DistanceMatrix
+        pairwise dissimilarity matrix.
+
+    Returns
+    -------
+    float:
+        the sum of all pairwise distances of dm for IDs in elements
+    '''
+
+    # some assertions
+    if type(dm) != DistanceMatrix:
+        raise TypeError('dm is not of type "DistanceMatrix"')
+    assert(len(elements) > 1)
+    assert(len(elements) <= len(dm.ids))
```

Comment:
shit, I don't know exactly what you are referring to.
Line 48: a Haskell habit. I like to check types, because it can save a lot of trouble, mainly because it clarifies expectations about the function.
Line 50: I could also return 0 as the sum of distances of only one point, but that would maybe confuse the user.
Line 51: the itertools.combinations operate on lists, i.e. duplicate elements are allowed. Thus, the sum for [A,B,C,A] would be larger than [A,B,C]